### PR TITLE
twirp.Erroer interface

### PR DESCRIPTION
**Pull request by marioizquierdo**
Issue #300

Allow errors returned by service handlers to implement a different interface `twirp.Erroer` to be serialized as Twirp errors.

The interface doesn't need to be exported, but having it exported documents the alternative better and makes the option more explicit.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by cespare** [comment]
Thanks for sending this PR! I will review this today.

**Comment by rhysh** [comment]
It's not clear to me that this new interface is required, or how the patterns it enables are different or better than what errors.As provides. I gave a specific example in https://github.com/twitchtv/twirp/issues/300#issuecomment-849826338.

File: protoc-gen-twirp/generator.go [hunk 1]
```diff
@@ -414,24 +416,37 @@ func (t *twirp) generateUtils() {
 	t.P(`}`)
 	t.P()
 
+	t.P(`// newTwirpError casts an error into the twirp.Error interface,`)
+	t.P(`// or calls .TwirpError() if it implements twirp.Erroer.`)
+	t.P(`// Otherwise it wraps the error as a twirp.Internal error.`)
+	t.P(`func newTwirpError(err error) `, t.pkgs["twirp"], `.Error {`)
```
[the remark this answers is not in the document]
Comment (marioizquierdo) [reply protoc-gen-twirp/generator.go:422]: Twirp tries to keep as much functionality as possible on the generated file, by design. That allows to re-generate and upgrade to newer versions without having to worry about version incompatibilities, which may become complex when importing multiple clients that were generated at different times. The less dependencies between the generated code and any other package, the better. Code size is not too bad at this point.

**Comment by marioizquierdo** [comment]
closing in favor of #323 323

**Pull request closed without merging**